lib/ddtrace/sampler.rb

```diff
@@ -42,8 +42,58 @@ def sample_rate=(sample_rate)
     end
 
     def sample(span)
-      span.sampled = ((span.trace_id * KNUTH_FACTOR) % Datadog::Span::MAX_ID) <= @sampling_id_threshold
       span.set_metric(SAMPLE_RATE_METRIC_KEY, @sample_rate)
+      span.sampled = ((span.trace_id * KNUTH_FACTOR) % Datadog::Span::MAX_ID) <= @sampling_id_threshold
+    end
+  end
+
+  # \RateByServiceSampler samples different services at different rates
+  class RateByServiceSampler < Sampler
+    def initialize(rate = 1.0, opts = {})
+      @env = opts.fetch(:env, Datadog.tracer.tags[:env])
+      @mutex = Mutex.new
+      @fallback = RateSampler.new(rate)
+      @sampler = {}
+    end
+
+    def sample(span)
+      key = key_for(span)
+
+      @mutex.synchronize do
+        @sampler.fetch(key, @fallback).sample(span)
+      end
+    end
+
+    def update(rate_by_service)
+      @mutex.synchronize do
+        @sampler.delete_if { |key, _| !rate_by_service.key?(key) }
+
+        rate_by_service.each do |key, rate|
+          @sampler[key] ||= RateSampler.new(rate)
+          @sampler[key].sample_rate = rate
+        end
+      end
+    end
+
+    private
+
+    def key_for(span)
+      "service:#{span.service},env:#{@env}"
+    end
+  end
+
+  # \PrioritySampler
+  class PrioritySampler
+    def initialize(opts = {})
+      @base_sampler = opts[:base_sampler] || RateSampler.new
+      @post_sampler = opts[:post_sampler] || RateByServiceSampler.new
+    end
+
+    def sample(span)
+      span.sampling_priority = 0
```

Comment:
Generally speaking, while writing the Python impl. we acknowledged, at some point, that it was easier and made more sense API-wise to have all the `sample()` family of methods return `true` or `false` and not modify the object in place. As in: https://github.com/DataDog/dd-trace-py/blob/master/ddtrace/sampler.py#L22 It's easier to test and the caller can make whatever decision on what to do, also, span becomes a read-only parameter, which has some advantages. This involves quite a rewriting of code at this step (because the existing sampler should be updated, not only the Priority one, but I think it's worth it). Might want to ask for @palazzem advice here.

Reply:
Humn, I guess I'll have to wait for @palazzem then, because I might be drifting too much from the Python implementation then.

I agree that the `#sample` return value should be a boolean, but I'd rather have a function with side effects here than breaking encapsulation by having different code paths in the caller for each concrete implementation of the `Sampler` – I think if we have a common API for samplers, we should leverage that and make them transparent to the caller.